File: solutions/java/circular-buffer/1/src/main/java/CircularBuffer.java
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class CircularBuffer<T> {
    private final List<T> buffer;
    private final int size;
    private int start;

    @SuppressWarnings("unchecked")
    public CircularBuffer(int size) {
        this.size = size;
        buffer = Arrays.asList((T[]) new Object[size]);
    }

    public T read() throws BufferIOException {
        T value = buffer.get(start);
        if (value == null) {
            throw new BufferIOException("Tried to read from empty buffer");
        }
        buffer.set(start, null);
        start = next(start);
        return value;
    }

    public void write(T value) throws BufferIOException {
        if (full()) throw new BufferIOException("Tried to write to full buffer");
        var i = start;
        while (buffer.get(i) != null) {
            i = next(i);
        }
        buffer.set(i, value);
    }

    public void clear() {
        buffer.replaceAll(value -> null);
    }

    public void overwrite(T value) throws BufferIOException {
        if (full()) {
            buffer.set(start, null);
            start = next(start);
        }
        write(value);
    }

    private int next(int i) {
        if (i < size - 1) return i + 1;
        else return 0;
    }

    private boolean full() {
        return buffer.stream().noneMatch(Objects::isNull);
    }
}

File: solutions/kotlin/circular-buffer/1/src/main/kotlin/CircularBuffer.kt
import kotlin.concurrent.fixedRateTimer

class EmptyBufferException : Throwable()

class BufferFullException : Throwable()

class CircularBuffer<T>(val size: Int) {

    private val buffer: MutableList<T?> = MutableList(size) { null }
    private var start = 0

    fun read(): T {
        val value = buffer[start] ?: throw EmptyBufferException()
        buffer[start] = null
        start = next(start)
        return value
    }

    fun write(value: T) {
        if (full()) throw BufferFullException()
        var i = start
        while (buffer[i] != null) {
            i = next(i)
        }
        buffer[i] = value
    }

    fun overwrite(value: T) {
        if (full()) {
            buffer[start] = null
            start = next(start)
        }
        write(value)
    }

    fun clear() {
        buffer.indices.forEach { buffer[it] = null }
    }

    private fun next(i: Int): Int = if (i < size - 1) i + 1 else 0


    private fun full(): Boolean = buffer.none { it == null }
}
